packages/playwright/src/AxeCoreWrapper/DefaultAxeCoreWrapper.cs

```diff
@@ -1,108 +1,297 @@
-
-            return DeserializeResult(jsonObject);
-        }
-
-        private static AxeResult DeserializeResult(object jsonObject)
-        {
-            if (jsonObject is null)
-            {
-                throw new Exception();
-            }
-
-            JObject jObject = JObject.FromObject(jsonObject);
-
-            return new AxeResult(jObject);
-        }
-
-        private static List<AxeRuleMetadata> DeserializeRuleMetadata(object jsonObject)
-        {
-            string result = JsonConvert.SerializeObject(jsonObject);
-            var results = JsonConvert.DeserializeObject<List<AxeRuleMetadata>>(result);
-
-            if (results is null)
-            {
-                throw new Exception();
-            }
-
-            return results;
-        }
-    }
-}
+#nullable enable
+
+using Deque.AxeCore.Commons;
+using Deque.AxeCore.Playwright.AxeContent;
+using Microsoft.Playwright;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Deque.AxeCore.Playwright.AxeCoreWrapper
+{
+    /// <inheritdoc/>
+    internal class DefaultAxeCoreWrapper : IAxeCoreWrapper
+    {
+
+        private readonly IAxeContentEmbedder m_axeContentEmbedder;
+        private static readonly DefaultContractResolver camelCaseContractResolver = new DefaultContractResolver
+        {
+            NamingStrategy = new CamelCaseNamingStrategy()
+        };
+        private static readonly JsonSerializerSettings JsonSerializerSettingsFinishRun = new JsonSerializerSettings
+        {
+            ContractResolver = camelCaseContractResolver,
+            Formatting = Formatting.None,
+            NullValueHandling = NullValueHandling.Include
+        };
+
+        public DefaultAxeCoreWrapper(IAxeContentEmbedder axeContentEmbedder)
+        {
+            m_axeContentEmbedder = axeContentEmbedder;
+        }
+
+        /// <inheritdoc/>
+        public async Task<IList<AxeRuleMetadata>> GetRules(IPage page, IList<string>? tags = null)
+        {
+            await m_axeContentEmbedder.EmbedAxeCoreIntoPage(page, false);
+
+            object jsonObject = await page.EvaluateAsync<object>("(paramTags) => window.axe.getRules(paramTags)", tags);
+            return DeserializeRuleMetadata(jsonObject);
+        }
+
+        /// <inheritdoc/>
+        public async Task<AxeResult> Run(IPage page, AxeRunContext? context = null, AxeRunOptions? options = null)
+        {
+            var rawContextArg = JsonConvert.SerializeObject(context);
+            AxeResult axeResult = await RunInner(page.MainFrame, rawContextArg, options);
+
+            return axeResult;
+        }
+
+        /// <inheritdoc/>
+        public async Task<AxeResult> RunOnLocator(ILocator locator, AxeRunOptions? options = null)
+        {
+            var frame = await (await locator.ElementHandleAsync()).OwnerFrameAsync();

```

Comment:
I think this new behavior of using the locator's owner frame instead of the page is good, but it looks like there's some missing test coverage for the new functionality - ideally there'd be some cases covering locators in iframes and/or shadow DOMs.